### ISSUE TITLE
[BUG] What is Channel unknown

### ISSUE THREAD

**Issue by tazik561** [leoafarias/fvm#305]
I am using fvm to manage flutter sdk , I installed sdk 2.0.6 by fvm. 
And I when I run fvm releases, It's result is:
```
Mar 19 21  │ 2.0.3            
Apr 2 21   │ 2.0.4            
Apr 15 21  │ 2.2.0-10.1.pre   
Apr 15 21  │ 2.2.0-10.1.pre   
Apr 16 21  │ 2.0.5            
Apr 27 21  │ 2.3.0-0.1.pre    
Apr 29 21  │ 2.2.0-10.2.pre   
Apr 30 21  │ 2.0.6            
May 10 21  │ 2.2.0-10.3.pre   
May 10 21  │ 2.3.0-1.0.pre    
May 18 21  │ 2.2.0            
May 18 21  │ 2.3.0-12.1.pre   
--------------------------------------
May 19 21  │ 2.2.0             beta
--------------------------------------
--------------------------------------
May 27 21  │ 2.2.1             stable
--------------------------------------
--------------------------------------
May 28 21  │ 2.3.0-16.0.pre    dev
--------------------------------------
```
Which type of version is stable?  for example 2.0.6

Now when I run fvm flutter doctor I got this:
```
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel unknown, 2.0.6, on Linux, locale en_US.UTF-8)
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.3)
[✓] Chrome - develop for the web
[!] Android Studio (not installed)
[✓] VS Code (version 1.56.2)
[✓] Connected device (2 available)
```

! Doctor found issues in 1 category.
 What is Channel unknown,?

I want just install stable version.

**Comment by leoafarias** [comment]
@tazik561 duplicate of this https://github.com/leoafarias/fvm/issues/291, to summarize there is a difference between a release channel and a release. A release can be part of multiple channels. This is correct behavior, and part of an ongoing discussion within the Flutter repo. https://github.com/flutter/flutter/issues/40950

**Comment by tazik561** [comment]
> @tazik561 duplicate of this #291, to summarize there is a difference between a release channel and a release. A release can be part of multiple channels. This is correct behavior, and part of an ongoing discussion within the Flutter repo. [flutter/flutter#40950](https://github.com/flutter/flutter/issues/40950)

Now , 2.0.6 is stable? and can I change channel unknown?

**Comment by leoafarias** [comment]
The only way the channel is known is if you install a release channel. Then flutter tooling will know that the tag is part of a release channel. That means you would have to do `fvm use stable`, and stay in that channel with its upgrades.

**Comment by leoafarias** [comment]
@tazik561 are you having a particular issue with the fact that it displays unknown?

**Comment by tazik561** [comment]
> The only way the channel is known is if you install a release channel. Then flutter tooling will know that the tag is part of a release channel. That means you would have to do `fvm use stable`, and stay in that channel with its upgrades.

Not possible because stable channel is  `2.2.1` and my project is not support this version yet.

**Comment by tazik561** [comment]
> > The only way the channel is known is if you install a release channel. Then flutter tooling will know that the tag is part of a release channel. That means you would have to do `fvm use stable`, and stay in that channel with its upgrades.
> 
> Not possible because stable channel is `2.2.1` and my project is not support this version yet.

Yes, I lost hot reload in this version and when I doing hot reload my application is terminated.

**Comment by leoafarias** [comment]
@tazik561 In that case you need to pin the version directly. There doesn't seem to be any "issues" caused by having the channel unknown unless you try to do the upgrade workflow.

**Comment by tazik561** [comment]
Not possible because stable channel is 2.2.1 and my project is not support this version yet. Thanks

**Comment by leoafarias** [comment]
I understand. You can continue to pin the version directly `fvm use 2.0.6` that should work just fine. That is why I asked above, are you having issues because of the channel "unknown"?

**Comment by leoafarias** [comment]
Ok I will go ahead and close this.

**Comment by knaeckeKami** [comment]
See https://github.com/flutter/flutter/issues/84347#issuecomment-862743141

TL;DR:
Flutter uses the current channel to figure out what features to enable
See https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/features.dart#L367

When you use fvm, the channel is unknown, so the experimental features of master are enabled.
There was a bug that then caused hot reload to stop working.

So actually, the fact that the channel is unknown can cause issues.

**Comment by leoafarias** [comment]
@knaeckeKami thanks will check it out to see if there is a proper workaround.

**Comment by leoafarias** [comment]
@knaeckeKami and @kuhnroyal  if you guys can help test to make sure you are getting the correct behavior `pub global activate fvm 2.1.2-dev.0`,

the only downside is that the upgrade message shows up for the version and might cause users to upgrade.. I wonder if I should put a guard if version is not a channel.

**Comment by kuhnroyal** [comment]
The version default to stable now which helps with my current problem.

But I don't think this is enough. It needs to support version X on channel Y format somehow which means it needs to support 2 CLI arguments and have the optional channel in the JSON config.
If there is no channel defined it needs to default to `stable`.

**version and channel**
* install specific version from channel

**channel only**
* install latest version from channel

**version only**
* try to find most mature channel with the version and install - `stable` > `beta` > `dev`
* warn users if the cannel is not `stable`


A good starting point for this logic is https://github.com/subosito/flutter-action

This probably is a breaking change.

**Comment by kuhnroyal** [comment]
For example version 2.0.0 is available on `stable` and `beta`.
I don't have a way to install it form the `beta` channel atm.

**Comment by leoafarias** [comment]
Yes I wanted to avoid that but makes sense. I can pass a flag —channel if using with a version and prioritize that one.

Not sure if I should warn the users as I don’t think is common for people to install a stable version from beta. But we should allow the override.

**Comment by kuhnroyal** [comment]
> Not sure if I should warn the users as I don’t think is common for people to install a stable version from beta. But we should allow the override.

I would add a warning when there is only a version configured and this version is not on stable, basically telling the user which channel was detected and that they should add the `--channel` flag to hide this warning and make it explicit.

**Comment by knaeckeKami** [comment]
Yeah, with 2.1.2-dev.0 the channel is detected correctly and this fixes the issue with the experimental features to be enabled on a supposedly "stable" version.

I agree that there should probably be a way to pick the channel for a given version if it was released on multiple versions.

**Comment by leoafarias** [comment]
Any suggestions how to handle the upgrade? The users now will be presented with the message to upgrade.

once they do fvm will lose track of the version and it won’t match, I can’t do much since this can happen from outside of fvm.

should we just have a notice and maybe flag if you run “fvm flutter upgrade” on a non channel version?

**Comment by kuhnroyal** [comment]
> should we just have a notice and maybe flag if you run “fvm flutter upgrade” on a non release channel?

Yea I would intercept this case and make it an interactive question y/n and tell them the consequences. 

And maybe proxy all other commands with `--no-version-check`.

**Comment by leoafarias** [comment]
I will have to store that version also, in the config... seems like there is a lot of complexity for something that I am not sure is done on purpose. Having a stable version on a release, especially since they make it to beta after stable most of the time.

**Comment by knaeckeKami** [comment]
I didn't really look into the architecture of fvm so I don't know what is easy to implement, but maybe you could also filter to "upgrade" message from being printed?

Also fvm could have a check if the current flutter version matches the expected version.

And yeah `fvm flutter upgrade` should be flagged.

Honestly, maybe we should open an issue on flutter.

**Comment by kuhnroyal** [comment]
Linking this for reference https://github.com/flutter/flutter/issues/59533

**Comment by knaeckeKami** [comment]
Also I opened #84781, which would at least cause more awareness of that behaviour

**Comment by leoafarias** [comment]
Hey guys I think I found an elegant solution for the channel version problem. Right now the behavior will be.
- `beta` => Latest version on beta
- `2.2.2` => Finds channel on priority (stable, beta, dev)
- `beta-2.2.2` => Gets 2.2.2 from beta channel
- 'commit_hash` => Gets commit from master channel

This `channel_name+version` keeps all the rest of the config and tooling working how is expected. Here are the behavior below.

![CleanShot 2021-06-17 at 10 51 43](https://user-images.githubusercontent.com/435833/122421792-a3f9de80-cf5a-11eb-8a5b-725d1e22fa79.png)

**Comment by kuhnroyal** [comment]
This would be fine for me, I would personally go with `beta/2.2.2` as some versions already contains dashes.

**Comment by leoafarias** [comment]
@kuhnroyal wants to avoid headaches with `/` and the file system since this is the directory that will look for the cache version, including any custom versions. Now for example you can have any flutter fork inside any named directory and just call it.

**Comment by kuhnroyal** [comment]
What about `2.2.2@beta` - I think this is even better

**Comment by leoafarias** [comment]
`pub global activate fvm 2.1.2-dev.1` has the `--no-version-check` and also the new convention `2.2.2@beta`

**Comment by leoafarias** [comment]
2.2.0 release closing this

**Comment by kuhnroyal** [comment]
Getting this on `fvm flutter pub run`:

```
Could not find an option named "no-version-check".
``` 

Need to figure out which commands support this. Probably only plain flutter commands an not those that proxy themselves to dart/pub.

**Comment by leoafarias** [comment]
Hmmm…  I am adding the “—no-version-check” and seems it would also run on pub…  maybe I need to change to add to the first index of args instead of in the end .

I am not in front of the computer. But if you want to dia quick fix, it is on flutter_command.dart and I can merge

**Comment by leoafarias** [comment]
Went ahead and removed it for now until I have a chance to write testes around this
